lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java

```diff
@@ -277,31 +288,44 @@ public ScoreMode scoreMode() {
     return hitsThresholdChecker.scoreMode();
   }
 
+  public boolean isQueueFull() {
+    return pqTop != null && pqTop.score != Float.NEGATIVE_INFINITY;
+  }
+
   protected void updateMinCompetitiveScore(Scorable scorer) throws IOException {
-    if (hitsThresholdChecker.isThresholdReached()
-          && ((bottomValueChecker != null && bottomValueChecker.getBottomValue() > 0)
-          || (pqTop != null && pqTop.score != Float.NEGATIVE_INFINITY))) { // -Infinity is the score of sentinels
-      // since we tie-break on doc id and collect in doc id order, we can require
-      // the next float
-      float bottomScore = Float.NEGATIVE_INFINITY;
-
-      if (pqTop != null && pqTop.score != Float.NEGATIVE_INFINITY) {
-        bottomScore = Math.nextUp(pqTop.score);
-
-        if (bottomValueChecker != null) {
-          bottomValueChecker.updateThreadLocalBottomValue(pqTop.score);
+    if (hitsThresholdChecker.isThresholdReached()) {
+      boolean hasChanged = false;
+      if (isQueueFull()) {
+        // since we tie-break on doc id and collect in doc id order, we can require
+        // the next float
+        float localMinScore = Math.nextUp(pqTop.score);
+        if (localMinScore > minCompetitiveScore) {
+          hasChanged = true;
+          minCompetitiveScore = localMinScore;
+          if (bottomValueChecker != null) {
```

Comment:
I am not sure about the benefit of this refactoring : seems like it is still checking if the local scorer's minimum score can be improved by either pqTop's score or the global minimum score, which is what was being done earlier? Could you describe the idea behind this refactoring, please?

Reply:
I agree, this refactoring was made because I wanted to call `updateMinCompetitiveScore` eagerly on non-competitive hits so I added some checks to avoid propagating the same minimum score multiple times. It is no longer needed since I removed this pre-optimization in the last iteration. I pushed a new [commit](https://github.com/apache/lucene-solr/pull/904/commits/5895797faed1f32b1fd2b733de66ea9592bfac76) to restore the original one.